### PR TITLE
Split sampling module tracing, add penalties, seed 

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/32212)

needs vllm changes: https://github.com/tenstorrent/vllm/pull/252

This PR refactors the sampling module to support split tracing and adds a new penalties module for presence, frequency, and repetition penalties. The main changes reorganize the sampling code into a modular structure under models/common/sampling/ with a new SamplingGenerator facade that manages both TTSampling and the new TTPenalties modules.

Introduced SamplingGenerator class that provides a unified interface for sampling with optional trace capture
Added TTPenalties module to apply presence, frequency, and repetition penalties on-device
Extended SamplingParams dataclass to include penalty parameters
Updated model classes to use the new sampling interface

### Checklist
blackhole: https://github.com/tenstorrent/tt-metal/actions/runs/19730893878
galaxy: https://github.com/tenstorrent/tt-metal/actions/runs/19734825785
single card: https://github.com/tenstorrent/tt-metal/actions/runs/19734840618
t3k: https://github.com/tenstorrent/tt-metal/actions/runs/19734866395

### CODE REVIEW

**Comment by sraizada-tt** [comment]
@copilot can you review

File: models/common/sampling/tt_penalties.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+    # If token appears in prompt or output, apply, otherwise use 1.0 for no-op.
+
+    combined_mask = ttnn.add(context.prompt_mask, context.output_mask, **op_kwargs)
+    penalties = ttnn.where(combined_mask, context.repetition_penalties, 1.0, **op_kwargs)
+    inverse_penalties = ttnn.reciprocal(penalties, **op_kwargs)
+
+    # If logits are positive, divide by penalty, otherwise multiply by penalty.
+
+    scaling = ttnn.where(ttnn.gt(logits, 1), inverse_penalties, penalties, **op_kwargs)
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/common/sampling/tt_penalties.py]: @sraizada-tt is this accurate? Why only greater than 1 and not 0?

File: models/common/sampling/tt_penalties.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    """
+    Penalty module with persistent device tensors, similar to TTSampling.
+    """
+
+    def __init__(self, mesh_device, max_batch_size: int, vocab_size: int, sub_core_grids=None):
+        super().__init__()
+        self.mesh_device = mesh_device
+        self.cluster_shape = mesh_device.shape
+        self.max_batch_size = 32  # max_batch_size
+        self.vocab_size = 128 * 1024  # vocab_size
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/common/sampling/tt_penalties.py]: Agree with this. 

The vocab size should always be given by the model params as it differs.
The max_batch size we currently don't have this information stored on a per-model-basis, but we'll need to accomodate higher batch sizes as we move along. Maybe add a TODO comment for the batch size on why we set to 32 and mentioning that it needs to be expanded at some point.

File: models/common/sampling/tt_penalties.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        self.token_bin_counts_and_mask(
+            new_tokens=new_tokens,
+            counts=self.output_counts_gathered,
+            counts_sliced=self.output_counts,
+            mask=self.output_mask,
+        )
+
+    def token_bin_counts_and_mask(self, new_tokens, mask, counts=None, counts_sliced=None):
+        # counts_new = ttnn.scatter_add(1, new_tokens)
+        # fallback
```
Comment (mtairum) [inline models/common/sampling/tt_penalties.py]: Please add TODO with the Open issue that will address this fallback.

File: models/common/sampling/tt_penalties.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            sub_core_grids=self.sub_core_grids,
+        )
+        tt_logits = ttnn.typecast(tt_logits, ttnn.bfloat16)
+        tt_logits = ttnn.pad(
+            tt_logits, [(0, 0), (0, 0), (0, 0), (0, self.vocab_size // self.cluster_shape[-1] - tt_logits.shape[-1])], 0
+        )
+
+        reshaped = ttnn.reshape(tt_logits, (-1, tt_logits.shape[-1]))
+
+        apply_penalties(reshaped, context)
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/common/sampling/tt_penalties.py:246]: Yes please.

File: models/common/sampling/tt_penalties.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+class TTPenalties(LightweightModule):
+    """
+    Penalty module with persistent device tensors, similar to TTSampling.
+    """
+
+    def __init__(self, mesh_device, args):
+        super().__init__()
+        self.mesh_device = mesh_device
+        self.cluster_shape = mesh_device.shape
+        self.max_batch_size = 32  # max_batch_size -- penalties and sampling only run for padded batch size
```
Comment (mtairum) [inline models/common/sampling/tt_penalties.py:81]: Should we track potentially adding support for other batch sizes at a later date?
Thinking about MoE models and running batch sizes > 32.

**Review by mtairum — APPROVED** [review]
Approving to unblock. 

Let's make sure we test all these new sampling features either in unit tests or full model demo.

File: models/common/sampling/tt_penalties.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/sampling/tt_penalties.py:246]: it is inplace

File: models/common/sampling/tt_penalties.py (hunk 5 above)
Comment (sraizada-tt) [reply models/common/sampling/tt_penalties.py:81]: the sampling op only supports 32 users right now. 
adding support for more users is non-trivial.
worst case we can loop in batches of 32 for sampling

**Review by djordje-tt — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 6]
```diff
@@ -577,7 +608,15 @@ def _capture_decode_trace_text(
             device_inputs_i = copy_host_to_device(host_inputs, mesh_device=self.model_args[i].mesh_device)
             device_inputs.append(device_inputs_i)
 
+        split_sampling_per_device = []
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Can remove unused list. Also what does split_enabled mean? Can that be renamed to capture_sampling_trace? Same for split_sampling_enabled above.

File: models/tt_transformers/tt/generator.py [hunk 7]
```diff
@@ -444,39 +463,51 @@ def decode_forward_text(
...
                     sampling_params_list[i], 32
                 )  # Sampling needs params padded to 32 regardless of batch_size
-                self.model[i].tt_sampling.reset_params(
-                    k=formatted_params.top_k,
-                    p=formatted_params.top_p,
-                    temp=formatted_params.temperature,
-                )
+                sampling_module = getattr(self.model[i], "sampling", None)
+                if sampling_module is None:
+                    continue
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Should there be an assert here since sampling_params is provided by sampling_module doesn't exist?

File: models/tt_transformers/tt/generator.py (hunk 6 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: the sampling trace is captured anyway -- what that variable indicates is whether we want to split the decoder trace into two parts upto logits and then, sampling -- added comment

File: models/tt_transformers/tt/generator.py (hunk 7 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: done

**Review by bbradelTT — APPROVED** [review]

**Review by skhorasganiTT — APPROVED** [review]
